### DISCUSSION TITLE
How to make relationships on User model?

### DISCUSSION THREAD

**Discussion by PelicanQ** [fastapi-users/fastapi-users#1338] [Q&A]
My register route fails as soon as User has a relationship attribute.

I have a User which should have many Post's. If I define my UserRead Pydantic model/schema to include posts, FastAPI will try to load posts for the response. This fails since FastAPI Users uses async SQLAlchemy which cannot lazy load (no implicit IO). Specifically I am calling the `/auth/register` as when following the [docs](https://fastapi-users.github.io/fastapi-users/12.1/configuration/full-example/).

These are my SQLAlchemy models:

```py
class Mixin(MappedAsDataclass, SQLAlchemyBaseUserTable[int]):
    pass

class User_DB(BaseModel_DB, Mixin):
    __tablename__ = "user_table"
    id: Mapped[int] = mapped_column(primary_key=True, init=False)  # type: ignore . It's fine
    firstname: Mapped[str] = mapped_column()
    posts: Mapped[list["Post_DB"]] = relationship(init=False)
```

```py
class Post_DB(BaseModel_DB):
    __tablename__ = "post_table"

    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    name: Mapped[str] = mapped_column()
    user_id: Mapped[int] = mapped_column(ForeignKey("user_table.id"))
```
and UserRead schema:

```py
class UserRead(fastapi_users_schemas.BaseUser[int], BaseSchema):
    firstname: str
    email: str
    posts: list[PostRead]  # uncommenting this makes the route work
```

The following is raised upon calling POST `/auth/register` with data. This error is from an async SQLAlchemy load which then tries to lazily access an attribute which is not eagerly loaded. Thus excluding posts from UserRead, the framework won't try loading it.

`Error extracting attribute: MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here. Was IO attempted in an unexpected place?`

So one possibility might be to specify eager load for every relationship on `User_DB` but why wouldn't the docs of FastAPI Users say this if it was the only way to make relationships work? It also forces eager loading on every relationship attribute we would wanna return.

**Q:** How are relationships meant to be used with FastAPI Users? Thank you!

The included routes don't even work with a simple related model. Reproduce with this [minimal repo I made](https://github.com/PelicanQ/fastapi-users-question).

**Comment by hgalytoby** [comment]
This is one of the solutions.

https://github.com/PelicanQ/fastapi-users-question/blob/main/db_models/user_model.py#L22

```py
 posts: Mapped[list["Post_DB"]] = relationship(init=False, lazy='selectin')
```


I'm not familiar with sqlalchemy 1.4 or 2.0, but I see from your error `Error extracting attribute: MissingGreenlet: greenlet_spawn has not been called; can't call await_only() here.`.

I understand that I should use `selectin` or `joined` from sqlalchemy 1.4, but I don't know how to write it in sqlalchemy 2.0 and above.

So I threw this question into google and looked for stack overflow to find the answer.

I don't think it's too hard to find the answer to this problem.

You should learn how to do the sqlalchemy relationship or you'll run into the same kind of problem when accessing data in a relationship.

**Reply by frankie567** [reply]
@hgalytoby is completely correct. More technical details about why it happens:

In a classic non-async context, when you do `user.posts` without having specified anything about eager loading, SQLALchemy will **implicitly** execute a new SQL query. It means that just writing `user.posts` will actually trigger blocking I/O. That's lazy loading.

In FastAPI Users, we use an *async driver* for performing I/O to the database. It means that every time you load data, it'll be something you should `await`. That's why it's not possible to do the implicit lazy loading like `user.posts`, and why you get the `greenlet_spawn` error. It's **not specific** to FastAPI Users, but rather to SQLAlchemy + Async.

If you want to load relationships, you have to tell SQLAlchemy upfront so it can do so in a single I/O roundtrip. In a way, I personally find it to be a more sane approach because your database is never reached implicitly. I used to work on big Django applications and lazy loading was a very classic performance issue.

---

So two solutions for you:

1. As @hgalytoby suggested, you can eager load your relationship with the `lazy` flag. Ref: https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#joined-eager-loading
2. Avoid to embed posts directly in the `/users/` routes. In a standard REST API, you would rather have a dedicated `/posts/` router to load the posts, and potentially have a filter to get the posts of a specific user.

**Reply by hgalytoby** [reply]
Thank you for the detailed explanation, and I agree that SQLAlchemy should be explicitly told to load associative data in a single operation instead of implicitly accessing it. This way you can avoid some APIs to get unnecessary associative data during operation, and thus improve performance.

The solutions you mentioned are excellent, and I especially like the second one.
